src/util/rocksdb/event_listener.rs

```diff
@@ -52,3 +57,127 @@ impl rocksdb::EventListener for EventListener {
+    pub cf: String,
+    pub output_level: i32,
+    pub total_input_bytes: u64,
+    pub total_output_bytes: u64,
+    pub start_key: Vec<u8>,
+    pub end_key: Vec<u8>,
+    pub input_props: Vec<SizeProperties>,
+    pub output_props: Vec<SizeProperties>,
+}
+
+impl CompactedEvent {
+    pub fn new(
+        info: &CompactionJobInfo,
+        start_key: Vec<u8>,
+        end_key: Vec<u8>,
+        input_props: Vec<SizeProperties>,
+        output_props: Vec<SizeProperties>,
+    ) -> CompactedEvent {
+        CompactedEvent {
+            cf: info.cf_name().to_owned(),
+            output_level: info.output_level(),
+            total_input_bytes: info.total_input_bytes(),
+            total_output_bytes: info.total_output_bytes(),
+            start_key: start_key,
+            end_key: end_key,
+            input_props: input_props,
+            output_props: output_props,
+        }
+    }
+}
+
+pub type Filter = fn(&CompactionJobInfo) -> bool;
+
+pub struct CompactionListener {
+    ch: Box<Fn(CompactedEvent) + Send + Sync>,
+    filter: Option<Filter>,
+}
+
+impl CompactionListener {
+    pub fn new(
+        ch: Box<Fn(CompactedEvent) + Send + Sync>,
+        filter: Option<Filter>,
+    ) -> CompactionListener {
+        CompactionListener {
+            ch: ch,
+            filter: filter,
+        }
+    }
+}
+
+impl rocksdb::EventListener for CompactionListener {
+    fn on_compaction_completed(&self, info: &CompactionJobInfo) {
+        if let Some(ref f) = self.filter {
+            if !f(info) {
+                return;
+            }
+        }
+
+        let mut input_files = HashSet::default();
+        let mut output_files = HashSet::default();
+        for i in 0..info.input_file_count() {
+            info.input_file_at(i)
+                .to_str()
+                .map(|x| input_files.insert(x.to_owned()));
+        }
+        for i in 0..info.output_file_count() {
+            info.output_file_at(i)
+                .to_str()
+                .map(|x| output_files.insert(x.to_owned()));
+        }
+        let mut input_props = Vec::with_capacity(info.input_file_count());
+        let mut output_props = Vec::with_capacity(info.output_file_count());
+        let iter = info.table_properties().into_iter();
+        for (file, properties) in iter {
+            if let Ok(prop) = SizeProperties::decode(properties.user_collected_properties()) {
+                if input_files.contains(file) {
+                    input_props.push(prop);
+                } else if output_files.contains(file) {
+                    output_props.push(prop);
+                }
+            } else {
+                warn!("Decode size properties from sst file failed.");
+                return;
+            }
+        }
+
+        if input_props.is_empty() && output_props.is_empty() {
+            return;
+        }
+
+        let mut smallest_key = None;
+        let mut largest_key = None;
+        for prop in &input_props {
```

Comment:
can we meet the case that inputs are `[a, b), [c, d)`, so we also need to care unrelated `[b, c)`?

Reply:
It doesn't matter, we will check the decline bytes for each covered region.